tripal/tripal
# Optimization of Publishing Entities

Issue #530 

## Type(s) of Change(s)


## Description


This PR optimizes the performance of Publishing Tripal Content. Specifically, I see a performance increase from 2.6 minutes to 1.17 minutes to publish 500 entities. Furthermore, publishing no longer grinds my site to a halt but instead there seems to be barely an impact to users.

Optimizations done:
- The bundle  and term objects were being loaded (4 and 1 time(s) respectively) per entity published. This is now cached by loading it once at the beginning of the publish job and then passing it to $entity->create(), $entity->save(), $entity->setAlias(), and $entity->setTitle() through a generic $cache variable. In all cases, I still load the bundle/term if it's not set in the cache making this a backwards compatible change.
- ~~There were nested transactions (publish and $entity->save) which caused unnecessary overhead. Instead I used the same $cache variable to pass the transaction through the save process.~~
- The Drupal Field cache is cleared for each entity saved to ensure new values can be loaded. However, during publish we don't attach fields so I added a conditional to ensure this cache isn't cleared during publish.
- I chunked the publishing process into 500 entity chunks to drastically speed up the select query, limit transaction time and allow reporting of progress to the jobs UI.

Additional Note: There was a comment in $entity->setAlias() implying `SELECT DISTINCT SUBSTRING_INDEX(source, '/', 1) AS path FROM url_alias` was causing performance issues during publish. I tracked this query down to drupal_path_alias_whitelist_rebuild() but didn't see it called during publish personally. We may want to look into this further.

## Testing?




1. Navigate to Admin > Tripal Content > Publish Tripal Content
2. Select a Tripal Content Type you have unpublished content for and submit the publishing job.
3. Check that the runtime reported is reasonable. NOTE: It may not be exactly the same as mine since runtime is obviously hardware dependent. Ideally you would have published the same set and timed it without the PR to allow you to compare the numbers.
4. Check that your site doesn't experience performance issues during the publish job.
5. Check that the entities published were created correctly by viewing one from Admin > Tripal Content.

## Additional Notes (if any):
